examples/metrics/projection.py

```diff
@@ -0,0 +1,382 @@
+=============================
+
+Projecting data onto P3 definitions.
+
+The P3 Analysis Library expects data to be prepared in a specific
+:ref:`format <performance_data>`. This format was inspired by the
+:ref:`terminology <terminology>` first introduced in "`Implications of a Metric
+for Performance Portability`_":
+
+.. _Implications of a Metric for Performance Portability:
+   https://doi.org/10.1016/j.future.2017.08.007
+
+**Problem**
+  A task with a pass/fail metric for which quantitative performance may be
+  measured.
+
+**Application**
+  Software capable of solving a *problem* with measurable correctness and
+  performance.
+
+**Platform**
+  A collection of software and hardware on which an *application* may run a
+  *problem*.
+
+These definitions are flexible, allowing the same performance data to be used
+for multiple case studies with different interpretations of these terms.
+
+Rather than store raw performance data in columns corresponding to these
+definitions, the P3 Analysis Library provides functionality to *project* raw
+performance data onto specific meanings of "problem", "application" and
+"platform".
+
+Using Projection to Rename Columns
+----------------------------------
+
+The simplest example of projection is a straightforward renaming of columns.
+
+Let's assume that we've collected some performance data from a few different
+implementations of a function, running a number of problem sizes on multiple
+machines.
+
+.. important::
+    Although we are looking at "function" performance here, the concepts
+    generalize to entire software packages.
+
+Our raw performance data might look like this:
+
+.. list-table::
+    :widths: 20 20 20 20
+    :header-rows: 1
+
+    * - size
+      - implementation
+      - machine
+      - fom
+
+    * - 128x128x128
+      - Library 1
+      - Cluster 1
+      - 0.5
+
+    * - 256x256x256
+      - Library 1
+      - Cluster 1
+      - 2.0
+
+    * - 128x128x128
+      - Library 2
+      - Cluster 1
+      - 0.7
+
+    * - 256x256x256
+      - Library 2
+      - Cluster 1
+      - 2.1
+
+    * - 128x128x128
+      - Library 1
+      - Cluster 2
+      - 0.25
+
+    * - 256x256x256
+      - Library 1
+      - Cluster 2
+      - 1.0
+
+    * - 128x128x128
+      - Library 2
+      - Cluster 2
+      - 0.125
+
+    * - 256x256x256
+      - Library 2
+      - Cluster 2
+      - 0.5
+
+The most obvious projection of this data onto P3 definitions is as follows:
+
+- Each input size maps to a different problem, because each input
+  represents a different task to be solved, with its own expected answer to
+  validate against.
+
+- Each implementation maps to a different application, because each
+  library's implementation of the function produces a solution for a given
+  input with measurable performance and correctness.
+
+- Each machine maps to a different platform, because each cluster name
+  describes the combination of hardware and software used to run the
```

Comment:
Empasize hardware _and_ software

Reply:
Done in a7b01bbe448c85841f12c30bfe2a83e54eb1f2e0.